src/dfx/src/main.rs

```diff
@@ -170,12 +174,35 @@ fn print_error_and_diagnosis(err: Error, error_diagnosis: Diagnosis) {
 }
 
 fn main() {
-    let cli_opts = CliOpts::parse();
+    let mut args = std::env::args_os().collect::<Vec<OsString>>();
+    let mut error_diagnosis: Diagnosis = NULL_DIAGNOSIS;
+
+    if let Ok(em) = ExtensionManager::new(dfx_version(), false) {
```

Comment:
@ericswanson-dfinity this line makes `e2e/tests-dfx/usage_env.bash` fail on this line https://github.com/dfinity/sdk/blob/41dc038704bc24cd2deebdb77365602a86148d0b/e2e/tests-dfx/usage_env.bash#L26
because `ExtensionManager::new` calls `pub fn get_bin_cache(v: &str)` which calls `get_cache_root` & `get_bin_cache_root` and they always create cache dir if it doesn't exist
https://github.com/dfinity/sdk/blob/41dc038704bc24cd2deebdb77365602a86148d0b/src/dfx-core/src/config/cache.rs#L34-L35
https://github.com/dfinity/sdk/blob/41dc038704bc24cd2deebdb77365602a86148d0b/src/dfx-core/src/config/cache.rs#L47-L48

Could you please advise how to solve it? I've been thinking about these options:
- extend function signatures for `fn get_bin_cache/get_cache_root/get_bin_cache_root` with param `initialize: bool` which when set to `false` guards the calls to `crate::fs::create_dir_all` 
- create additional function `fn get_bin_cache_unitialized() -> PathBuf`
- create additional function `fn is_cache_installed(v: Version) -> bool`, and make `ExtensionManager::new` return `Err` if cache not present

Reply:
If I understand it, `dfx identity new` creates the empty cache dir (as will any dfx command), rather than `dfx new`.

The purpose of this test is to show that DFX_CONFIG_ROOT and DFX_CACHE_ROOT control the locations of identities and the cache, respectively.

The failing line in question is verifying a known initial state, in order to prove that `dfx new` creates the cache directory under DFX_CACHE_ROOT and does not create it under HOME.

I would suggest deleting the cache directory before that failing line, either with `dfx cache delete` or through other means.

Reply:
> dfx identity new creates the empty cache dir (as will any dfx command), rather than dfx new.

Yes, exactly. I thought this is not ok? with the current setup even `dfx --help` would create an empty cache dir

Reply:
Sure, how about see if you can find a way to skip all of this if there's no cache (or no extensions directory in the cache).

```
    if let Ok(em) = ExtensionManager::new(dfx_version(), false) {
        match em.installed_extensions_as_clap_commands() {
```